gossipsub

vyzo opened the pull request:
Implements the gossipsub protocol; see https://github.com/vyzo/gerbil-simsub for a high-level literate specification.

TODO:
- [X] tests tests tests!
- [X] subscription announce messages, sent by base `PubSub`, need to be reliable so that we accurately track peers in a topic

vyzo commented:
Added a small zoo of basic gossipsub tests, including a mixed mode test with floodsub peers.

vyzo commented:
made a small tweak -- sources that have not joined the mesh also emit gossip.

vyzo commented:
Removed a potentially harmful topic membership check for mesh peers; potential inconsistency if the ANNOUNCE was lost (or reordered after GRAFT on retry).

whyrusleeping commented on gossipsub.go:
```diff
@@ -0,0 +1,569 @@
+		if len(peers) < GossipSubDlo {
+			ineed := GossipSubD - len(peers)
+			plst := gs.getPeers(topic, ineed, func(p peer.ID) bool {
+				_, ok := peers[p]
```
this looks like we're filtering something, mind adding a comment on what we're filtering on?

vyzo replied:
we are filtering the peers that are not already in our peer list; i can add a comment to that extent.

whyrusleeping commented on gossipsub.go:
```diff
@@ -0,0 +1,569 @@
+	}
+}
+
+func (gs *GossipSubRouter) heartbeat() {
```
This function feels too long. Mind trying to break it up a little bit?

vyzo replied:
Sure i'll refactor a bit, although I would like to keep the main logic together.

vyzo replied:
I can easily refactor out the part that sends the GRAFT/PRUNE with coalescing, which is also incidental for the logic in the heartbeat.

vyzo replied:
done

vyzo commented on gossipsub.go:
```diff
@@ -0,0 +1,569 @@
+}
+
+func (gs *GossipSubRouter) heartbeatTimer() {
+	ticker := time.NewTicker(1 * time.Second)
```
Not necessarily.
It will send GRAFT/PRUNE if it needs to adjust the overlay, and schedule gossip for piggybacking.
If the gossip is not sent by the next heartbeat, then it will be flushed in its own messages.

vyzo commented:
Implemented retry of ANNOUNCE messages in pubsub, now that we have a test that exercises the relevant code paths.

whyrusleeping commented on gossipsub_test.go:
```diff
@@ -0,0 +1,857 @@
+	}
+
+	// wait for heartbeats to build mesh
+	time.Sleep(time.Second * 2)
```
Why do we always have to wait for heartbeats? Is it because we don't send out subscription notices immediately now?

vyzo replied:
Well, subscription notices are still sent immediately, just retried if they fail.

Now that I think of it, we can probably greatly reduce this delay in almost all the tests -- maybe down to say 100ms, just enough for announcements to go out. 

My rationale for this delay was to wait a couple of heartbeats to avoid interference from nodes who have joined but haven't seen any peer announcements yet. Also, I wanted to avoid interference from the overlay construction, but it should still be connected after the announcements get sent and nodes pick their peers.

vyzo replied:
Actually, there is a genuine concern that is an artifact of the concurrent Join from subscriptions.
If we subscribe all the nodes together, then they won't have any peer announcements when the do the Join, and they'll have to wait a heartbeat before they start adding peers to the mesh.
We can avoid this if we add a small (say 10ms) delay after each subscription.

whyrusleeping replied:
Hrm... I'm very skeptical of 'fixing' things by adding delays.

vyzo replied:
Also, in many of the tests the subsscriptions are created before connecting the network, which means that all nodes start empty and build the mesh purely in the heartbeat

paralin replied:
@vyzo It's really really bad practice to use delays like this, especially without a select statement to escape out of it if the context is canceled.

vyzo replied:
@paralin , this is just a test that _needs_ a delay -- and there is nothing to cancel the context so a select would be totally useless.

paralin replied:
@vyzo Gotcha, I notice now it's a test.

whyrusleeping commented:
Could we add some tests that check the number of messages sent, and maybe a way of tracking the overall efficiency of the implementation (like how many nodes received the same message from multiple peers) maybe in terms of bandwidth overhead? like, received 500 bytes for every 200 bytes of useful data at a message size of 100 bytes.

vyzo commented:
Hrm, these are tests i would like to have too -- but not sure they are really unit tests.
What would be the conditions for test failure?

whyrusleeping commented:
Yeah, they are definitely integrations tests. No need to write them as unit tests. We should run these tests for floodsub and for gossipsub and compare the results, and choose some failure threshold, i.e. gossipsub should not use more bandwidth than floodsub

vyzo commented:
We have developed a conflict, so I will resolve and rebase.

vyzo commented:
Rebased; also added a context done check that was missing in the announce retry goroutine.

vyzo commented:
The `TestGossipsubControlPiggyback` would occasionally hang in line 704 because of #69, so I added a fix for the issue.

ORBAT commented:
Hi folks! Any particular reason you went with a custom protocol instead of something built on Chord/Pastry/PolderCast etc?

vyzo commented:
hrm, seems like the fix lost the coverage for control piggybacking -- probably because of the slowdown with the `Errorf` logging.
I will downgrade that to `Infof`.

vyzo commented:
@ORBAT several reasons: simplicity of implementation, robustness, and perhaps most important of all backwards compatibility with floodsub so that we can easily deploy.

vyzo commented:
and control piggyback coverage is back, at least for `GRAFT`.

whyrusleeping commented:
@ORBAT It looks like PolderCast didnt make it into our pubsub research reading list: https://ipfs.io/ipfs/QmNinWNHd287finciBwbgovkAqEBQKvnys1W26sY8uupc5/pubsub%20reading%20list.pdf
Likely because the paper costs $30 to read.

In any case, we've done a pretty thorough review of the problem space before we arrived at our version zero protocol, floodsub. With the idea that it is the base layer protocol, and provides very few guarantees. This code, which we're calling gossipsub is an iterative improvement over floodsub that essentially only adds a fairly simple tree pruning via gossip. Simplicity and ease of implementation are very important for us, gossipsub can be implemented in [150 lines of scheme](https://github.com/vyzo/gerbil-simsub/blob/master/simsub/gossip.ss) and not too many lines of go.

That said, this is still under review. Review of the protocol and/or implementation is very much welcome.

vyzo commented:
Rebased on master to absorb the cherry-pick.

whyrusleeping commented:
Review from while i was on a plane today:
1. docs please, its not clear that `Join` and `Leave` are just callbacks from the pubsub to the router
	- in general, we should have docs on the expectations and contracts that a router is expected to fulfil
2. why a random delay on announceRetry?
3. random thought, but what happens if some peers set their gossipsub parameters really low? like 1 or 2.
4. publishing to a topic you arent subscribed to seems like it might get expensive. also maybe a bit leaky? If i publish once to a topic, it looks like i set fanout peers for that topic. Does that ever get cleaned up?
5. any thought around getting DoSed through the 'topics' field filling up? There are probably a thousand other ways to DoS pubsub, but this one seems pretty easy
6. heartbeat interval should probably be configurable, with the default being a constant. Maybe link to your experiments showing why 1 second is a good default
7. i would turn the 'getNetSequenceNumber' thing in pubsub.Publish into a separate function
8. more docs on the different implications between mesh and fanout
9. in heartbeat, it seems like we could combine the looping over plst with the initial getPeers call. Why not just put the logic from that look in the closure youre passing?
10. is there no 'too many peers' check for fanout becoming too large?
11. in heartbeat, we call flush to push out piggybacked messages that didnt get sent, and then at the end of heartbeat we send some more messages. Why not call flush after that to potentially coalesce the grafts and prunes from this heartbeat with the gossip info in flush?

vyzo commented:
>   docs please, its not clear that Join and Leave are just callbacks from the pubsub to the router
>       in general, we should have docs on the expectations and contracts that a router is expected to fulfil

Sure, will add!

>    why a random delay on announceRetry?

I wanted to give the writer a chance to make progress, without an immediate retry. 
The delay could be fixed I guess, but it seemed better to randomize it as it avoids pathologies from periodic publish processes.

>    random thought, but what happens if some peers set their gossipsub parameters really low? like 1 or 2.

The overlay construction algorithm will continue to build connected overlays as long as `D-low >= 2` and `D-high >= D+1`. 
If you set all parameters to 1 or 2 globally then you will probably fail to connect non-trivially sized overlays.


>    publishing to a topic you arent subscribed to seems like it might get expensive. also maybe a bit leaky? If i publish once to a topic, it looks like i set fanout peers for that topic. Does that ever get cleaned up?

Well, it's not expensive as it is a passive operation (you select peers that have announced the topic).

But you are right, it _is_ leaky! -- if you only publish one message and then forget the topic, the fanout peers will linger.
I can add some logic in the heartbeat that cleans up `fanout` if you haven't published again for some time.

>    any thought around getting DoSed through the 'topics' field filling up? There are probably a thousand other ways to DoS pubsub, but this one seems pretty easy

What's the vector? Are you concerned about the `topics` field in the envelope causing excessive cpu usage from processing?

>    heartbeat interval should probably be configurable, with the default being a constant. Maybe link to your experiments showing why 1 second is a good default

We should probably make all parameters configurable, not just the heartbeat interval. 
The 1s choice was made with responsiveness in mind, and also to minimize the gossip linger interval. We can easily increase the heartbeat, but then
- it will take longer to adjust the overlay
- the gossip will linger longer, as messages are gossiped for 3 heartbeat cycles.

>    i would turn the 'getNetSequenceNumber' thing in pubsub.Publish into a separate function

sure, can do

>    more docs on the different implications between mesh and fanout

sure, i will add some comments

>    in heartbeat, it seems like we could combine the looping over plst with the initial getPeers call. Why not just put the logic from that look in the closure youre passing?

Not sure I follow. The closure is just a filter, it shouldn't have side-effects. 
Specifically, it is called for every peer to filter it and there is no guarantee that the peer will be selected,
so we can't fuse the subsequent iteration in the closure.

>    is there no 'too many peers' check for fanout becoming too large?

fanout cannot become too large, we only add peers if we have less than `D`.

>    in heartbeat, we call flush to push out piggybacked messages that didnt get sent, and then at the end of heartbeat we send some more messages. Why not call flush after that to potentially coalesce the grafts and prunes from this heartbeat with the gossip info in flush?

This may be a little cumbersome to write, as the flush should happen before we emit fresh gossip (otherwise it will flush the fresh gossip too, which is not what we want). 
I can break `emitGossip` out of the `graft/prune` loop, move the `sendGraftPrune` right after the iteration and then do flush, followed by another iteration over the topic lists to emit gossip.
Not sure if it's worth the trouble and code uglification though.

vyzo commented:
Updates based on the feedback:
- added docs for the router interface, the contract should be cleaner now
- made all gossipsub parameters, including the heartbeat interval, variables so that users can set their own configuration
- added expiration for fanout lists with a 1 minute TTL (thanks for spotting that leak!)
- refactored the next seqno contraption out of `pubsub.Publish`

vyzo commented:
@mgoelzer fixed in https://github.com/libp2p/go-floodsub/pull/67/commits/88a9bb054176be98655e0ea13294c20074b5c95b

Also a note about our review norms: we don't normally block merging with an `X` for minor things, we just inline comment and fix them.

vyzo commented on pubsub.go:
```diff
@@ -401,6 +442,12 @@ func msgID(pmsg *pb.Message) string {
+
 	if len(vals) > 0 {
 		// validation is asynchronous and globally throttled with the throttleValidate semaphore.
 		// the purpose of the global throttle is to bound the goncurrency possible from incoming
```
not a typo.

whyrusleeping replied:
not a typo?

vyzo replied:
more of a pun :)
goroutines provide goncurrency, just like coroutines provide concurrency.

vyzo replied:
obviously, we can fix it if you guys find my humor offensive :)

whyrusleeping replied:
lol, i don't mind

vyzo commented on pb/rpc.proto:
```diff
@@ -17,6 +19,30 @@ message Message {
 	repeated string topicIDs = 4;
 }
 
+message ControlMessage {
```
I am following the protocol style we use in other protobufs in ipfs; keep it simple, stupid.

lanzafame replied:
Totally understand, just unfortunate that we went with a style that isn't in line with what the tooling suggests.

prestonvanloon commented:
Is this still in review? Is there anything we can do to help out?

whyrusleeping commented:
@prestonvanloon I think we will merge it shortly, The only thing I want to hold off for is for @Stebalien to get back (he's out this week) and give a +1.

jamesray1 commented:
https://github.com/libp2p/go-floodsub/pull/79 should help for documentation.

jamesray1 commented on gossipsub.go:
```diff
@@ -0,0 +1,613 @@
+	peers   map[peer.ID]protocol.ID         // peer protocols
+	mesh    map[string]map[peer.ID]struct{} // topic meshes
+	fanout  map[string]map[peer.ID]struct{} // topic fanout
+	lastpub map[string]int64                // last pubish time for fanout topics
```
This comment should say publish. ;)

jamesray1 commented:
Could we have an ABI with JSON tests for other programming languages? This is needed for ensuring compatibility over the wire.

whyrusleeping commented:
> Could we have an ABI with JSON tests

What do you mean? Like test vectors and such for the wire protocol?

jamesray1 commented:
Yep, exactly! The goal is to make sure that if we have different implementations of gossipsub that they will be compatible.

whyrusleeping commented:
@jamesray1 I can add that to my todo list :)

In general though, it's all just that protobuf being sent back and forth.

jamesray1 commented:
Right, so if we use the same proto file we can make it compatible.

mhchia commented:
Will this PR be rebased to the current master?

It seems some of the APIs of the dependencies like `"github.com/libp2p/go-libp2p-netutil".GenSwarmNetwork` is changed.
So, I rebase this PR to the master branch locally. Some tests fail but I'm not sure the reason.
```
$ go test -run TestDenseGossipsub
--- FAIL: TestDenseGossipsub (1.03s)
	floodsub_test.go:53: dial attempt failed: <peer.ID X8bjX2> --> <peer.ID NtTuXy> dial attempt failed: EOF
FAIL
exit status 1
FAIL	github.com/libp2p/go-floodsub	1.049s
```
I changed the argument `d` passed to `denseConnect` from `10` to `8`, or other lower numbers to workaround this problem. But I didn't figure out the reason.
```
func denseConnect(t *testing.T, hosts []host.Host) {
	connectSome(t, hosts, 8)
}
```
Does anyone know the possible reason for this issue?  Or is this just the problem in my local environment?

Big thanks for reading my question.

whyrusleeping commented:
Hey @mhchia, sorry about that, We will get this rebased asap. I'll handle it in the morning if @Stebalien (who has more context on these changes) doesnt beat me to it.

Stebalien commented:
I've rebased but I'm having some trouble reproducing the issue. I'm currently running the test in a loop to see if that gets me anywhere.

Stebalien commented:
No dice.

jamesray1 commented:
@whyrusleeping I just read https://github.com/libp2p/interop/issues/1. Having a daemon will of course be useful, although not having to depend on code in Go is preferable, and JSON tests are needed.

whyrusleeping commented:
I had no reason not to merge this, so I did. Next steps, putting it into a flag in ipfs.

diasdavid commented:
@whyrusleeping shouldn't this be a separate pubsub implementation so that folks can pick the pubsub implementation to use? 

Will this PR make PubSub in go-ipfs not interop with js-ipfs?

diasdavid commented:
@whyrusleeping just confirmed that this package would be better named `go-pubsub` and then we would have two others for `go-floodsub` and `go-gossipsub` to plug in here but refactoring things in go is hard so that will happen later.

jamesray1 commented:
Also you really should use constant variables instead of literals.

mhchia commented:
@jamesray1 
Maybe it is because we can use different routing mechanism in the underlying overlay, not necessarily using DHT? In this case, "The initial contact nodes can be obtained via rendezvous with DHT provider records" might only be an example.

jamesray1 commented:
@mhchia, sure, that's fine.

whyrusleeping commented:
@diasdavid 

> but refactoring things in go is hard so that will happen later.

Its more 'extracting things into multiple packages in go is annoying to do when you might be changing things in both really soon'.

@jamesray1 @mhchia Yeah, the DHT is only an example. You can use any means to rendezvous. Take a look at our rendezvous spec proposal for ideas towards a more specialized way of doing rendezvous

jamesray1 commented:
What duration should we use for timeout requests? 

Context: implementing a system config for Kademlia to use to get nodes.

https://github.com/libp2p/rust-libp2p/blob/7507e0bfd9f11520f2d6291120f1b68d0afce80a/kad/src/high_level.rs#L36

As for the timeout duration, according to [RabbitMQ](https://www.rabbitmq.com/heartbeats.html), that is twice the heartbeat interval, which is 1 s in this Go implementation, therefore I think it should be 2 s. I am guessing to use [40000 s](https://github.com/amule-project/amule/blob/166ee417573e36c5162ad46b8da353d6b001ad03/src/include/protocol/ed2k/Constants.h#L34) for `kbuckets_timeout` (the Duration after which a node in the k-buckets needs to be pinged again.) but I'm not really sure, perhaps the spec and this implementation should also define initialization?

raulk commented on gossipsub.go:
```diff
@@ -0,0 +1,613 @@
+}
+
+func (gs *GossipSubRouter) handleIWant(ctl *pb.ControlMessage) []*pb.Message {
+	ihave := make(map[string]*pb.Message)
```
Does this need to be a map, or could it be a slice? Is it a map to deduplicate message IDs?

vyzo replied:
yes, it needs to deduplicate.
